virtcontainers/sandbox.go

```diff
@@ -1283,15 +1284,19 @@ func (s *Sandbox) CreateContainer(contConfig ContainerConfig) (VCContainer, erro
 		return nil, err
 	}
 
-	// Store it.
-	err = c.storeContainer()
+	// Sandbox is reponsable to update VM resources needed by Containers
+	s.updateResources()
 	if err != nil {
 		return nil, err
 	}
 
-	// Update sandbox config.
-	s.config.Containers = append(s.config.Containers, contConfig)
-	err = s.storage.storeSandboxResource(s.id, configFileType, *(s.config))
+	err = c.createContainer()
```

Comment:
I feel like we should call this before we update the VM resources.

Reply:
If we call it before the update VM resources, the resources wont exist at container creation, for some scenarios we should ensure they exist before create the container. For example, if the container cgroup has a cpuset information.
```diff
$ git diff
diff --git a/virtcontainers/sandbox.go b/virtcontainers/sandbox.go
index 85cbdcc..280ff58 100644
--- a/virtcontainers/sandbox.go
+++ b/virtcontainers/sandbox.go
@@ -1284,13 +1284,13 @@ func (s *Sandbox) CreateContainer(contConfig ContainerConfig) (VCContainer, erro                                               
                return nil, err
        }

-       // Sandbox is reponsable to update VM resources needed by Containers                                                                           
-       s.updateResources()
+       err = c.create()
        if err != nil {
                return nil, err
        }

-       err = c.create()
+       // Sandbox is reponsable to update VM resources needed by Containers                                                                           
+       s.updateResources()
        if err != nil {
                return nil, err
        }
```
Create a container with cpuset data, the agent will try to apply cgroups but vcpus resources are not created yet.
```bash
$  docker run -ti --runtime kata-runtime --rm --name kata-1 --cpus --cpuset-cpus="0-2" stress-ng  nproc                 
invalid argument "--cpuset-cpus=0-2" for "--cpus" flag: failed to parse --cpuset-cpus=0-2 as a rational number                                         
See 'docker run --help'.
$  docker run -ti --runtime kata-runtime --rm --name kata-1 --cpus 4 --cpuset-cpus="0-2" stress-ng  nproc               
docker: Error response from daemon: OCI runtime create failed: rpc error: code = Internal desc = Could not run process: container_linux.go:348: starting container process caused "process_linux.go:279: applying cgroup configuration for process caused \"failed to write 0-2 to cpuset.cpus: write /sys/fs/cgroup/cpuset/docker/051d8204fbb1d4585fff45b70fb8de303745070f39e9c267725abcd5a1dd2a71/cpuset.cpus: invalid argument\"": unknown.                         
```
With the current code this work

```bash
$ git reset --hard
$ make & make install
=$  docker run -ti --runtime kata-runtime --rm --name kata-1 --cpus 4 --cpuset-cpus="0-2" stress-ng  nproc               
3
```